### PR TITLE
persist: state pubsub design doc

### PR DESCRIPTION
**Pull request by pH14**
### Tips for reviewer

[Rendered](https://github.com/pH14/materialize/blob/persist-state-pubsub-design-doc/doc/developer/design/20230418_persist_state_pubsub.md).

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by danhhz — APPROVED** [review]
lgtm \o/!

all my comments are relatively minor

File: doc/developer/design/20230418_persist_push_based_updates.md [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+We will introduce several metrics to understand the behavior of the change:
+
+* # of updates pushed
+* # of updates received that apply cleanly & uncleanly
+* How frequently `Listen`s need to fallback to polling
+* End-to-end latency of appending data to observing it
```
[the remark this answers is not in the document]
Comment (pH14) [reply doc/developer/design/20230418_persist_push_based_updates.md:177]: was thinking of tagging messages with a CaS commit timestamp and plumbing that through. it's imperfect because clock skew etc but could be a decent ballpark

[the remark this answers is not in the document]
Comment (danhhz) [reply doc/developer/design/20230418_persist_push_based_updates.md:169]: we already put the timestamp on diffs! :D

and yeah, that seems fine. i was mostly curious if you'd come up with something clever to account for the clock offsets

**Comment by pH14** [comment]
cc @alex-hunt-materialize tagging you in here as a representative from Cloud, feel free to reassign. As the design is currently described, the questions we'd love input on from on the orchestration side: is work needed for `clusterd` to initiate connections to `envd`? any concerns with opening another port on `envd`?

File: doc/developer/design/20230418_persist_push_based_updates.md [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+directly instead. While we will continue to have these fallbacks, we anticipate message delivery being effective enough
+to eliminate the need for nearly all Cockroach reads.
+
+# Reference Explanation
+
+As a first implementation of the Persist sidechannel, we propose building a new RPC service that flows through `environmentd`.
+The service would be backed by gRPC with a minimal surface area -- to start, we would use a simple bidirection stream that
+allows each end to push updates back and forth, with `environmentd` broadcasting out any changes it receives.
+
+## RPC Service
```
Comment (bkirwi) [inline doc/developer/design/20230418_persist_push_based_updates.md:105]: This is maybe a bit too cute, but we could also GRPCify the existing `Consensus` interface, and use a "long polling" model for `scan` (where we block for some timeout before returning an empty list of changes, and if a relevant `compare_and_set` comes in in the meantime we resolve any pending `scan`s)  instead of push-based updates.

(I think that description might be too terse, but if it seems intriguing happy to discuss sync.)

**Review by bkirwi — COMMENTED** [review]
General summary for me:
- Enthusiastic about taking more advantage of local information instead of pushing everything through consensus.
- Nervous about pushing all changes through `environmentd`. That service is on the hot path for our most critical, latency-sensitive tasks; it's getting an increasing number of things like this that just need to run _somewhere_; conflicting requirements between those two needs are causing operational problems fairly regularly; this one seems to benefit very little from being run in that single process. (I do appreciate all the reasons why piggybacking `environmentd` is nice, though!)
- Not totally sold on the cost/benefit. In particular, while it's very easy to quantify consensus costs, I'm not sure how to quantify the cost of doing extra work in every `environmentd`.

File: doc/developer/design/20230418_persist_push_based_updates.md [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+directly, creating a mesh. In this model, `environmentd` would no longer need to broadcast out changes -- each process
+would send diffs directly to the processes that need them.
+
+This approach is additionally interesting in that it would set Persist up to communicate other types of data as well,
+not just pub-sub of diffs. One hypothetical -- shard writers could cache recently-written blobs to S3 in memory or on
+disk, and other processes could request those blobs from the writer directly, rather than paying the costs of going to
+S3, which is likely to have much longer tail latencies.
+
+The downsides of this approach include the complexity overhead, as it introduces the novel pattern of intra-`clusterd`
+communication in a way that does not exist today, and the scaling limitations imposed by all-to-all communication.
```
Comment (alex-hunt-materialize) [inline doc/developer/design/20230418_persist_push_based_updates.md:237]: Besides an increased number of connections and opened sockets, what are the scaling limitations?

**Review by alex-hunt-materialize — APPROVED** [review]
> cc @alex-hunt-materialize tagging you in here as a representative from Cloud, feel free to reassign. As the design is currently described, the questions we'd love input on from on the orchestration side: is work needed for `clusterd` to initiate connections to `envd`? any concerns with opening another port on `envd`?

I don't think any work is needed on the cloud side for `clusterd` to initiate connections to `environmentd`. We already have a network policy allowing all traffic between pods in the same environment. I've got no concern about opening another port on `environmentd` either.


I do wonder what the bandwidth and CPU load increases for `environmentd` are, but I've got no context for how much this matters.

I also wonder how this design will fare when we eventually have multiple `environmentd` instances running at once, both with a single leader as described in https://github.com/MaterializeInc/cloud/pull/6165 and in the hypothetical far future of true `environmentd` HA.

File: doc/developer/design/20230418_persist_push_based_updates.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+## Lifecycle
+[lifecycle]: #lifecycle
+
+We will use a feature flag that can quickly toggle on the publication / subscription to updates dynamically. Our metrics
+dashboards will inform our rollout process.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Looking forward, we know our polling-based discovery is not long-term sustainable and will need replacement. Given that,
```
[the remark this answers is not in the document]
Comment (pH14) [reply doc/developer/design/20230418_persist_push_based_updates.md:184]: Good point, I didn't lay out the argument for why it's not sustainable here. Some thoughts - 

Reads scale less efficiently than writes - because polling read loops are n:1 with tables/sources, their cost multiplies more quickly with customer usage than for writes. For an empty env, its impact on CRDB CPU will be something like 80:20 writes vs read load, but for our biggest env in prod, it's more like 45:55 -- the cost of reads increases as users push their environment harder, and in fact a full 10% of our entire CRDB CPU budget in us-east-1 prod is polling reads for a single customer right now.

We could argue that this cost is just part of doing business, which is a fair argument, but since we're looking to trim CRDB load / infra cost, the cost of consensus/durably recording Persist state changes feels fundamental whereas the cost of "discovery" of that information could be made much more efficient.

Having said all that, maybe what I was actually looking to say here is "financially unsustainable" rather than "technically unsustainable", because I'm sure CRDB backed by an infinite budget could scale as far as we need it to go to handle polling reads indefinitely

File: doc/developer/design/20230418_persist_push_based_updates.md (hunk 2 above)
Comment (pH14) [reply doc/developer/design/20230418_persist_push_based_updates.md:105]: 👍 let's sync up through some medium that is more responsive than GH comments

**Comment by pH14** [comment]
> I do wonder what the bandwidth and CPU load increases for environmentd are, but I've got no context for how much this matters.

My gut says both are low until we get huge #s of shards, but without trying it out on a real env it's hard to know for sure. We'll definitely rethink the approach if the initial test envs show it's an overly expensive process

> I also wonder how this design will fare when we eventually have multiple environmentd instances running at once, both with a single leader as described in https://github.com/MaterializeInc/cloud/pull/6165 and in the hypothetical far future of true environmentd HA.

Ooo yeah that's an excellent point. I think it works if the `clusterd` get a connection address that points to the generation-namespaced `envd` that created them, similar to how `envd` only connects to the `clusterd` it created and we disallow cross-version/update connections. But, will need to ponder that more...

**Comment by pH14** [comment]
>Nervous about pushing all changes through environmentd. That service is on the hot path for our most critical, latency-sensitive tasks; it's getting an increasing number of things like this that just need to run somewhere; conflicting requirements between those two needs are causing operational problems fairly regularly; this one seems to benefit very little from being run in that single process. (I do appreciate all the reasons why piggybacking environmentd is nice, though!)

> Not totally sold on the cost/benefit. In particular, while it's very easy to quantify consensus costs, I'm not sure how to quantify the cost of doing extra work in every environmentd.

Yep totally agree, putting the routing in `envd` definitely concerns me. OOC what would you like to see here to vet whether or not the added cost to `envd` is justified?

I'll use this as an opportunity to tag in @mjibson as a representative from Surfaces (feel free to reassign), since I imagine Surfaces to be most acutely aware of `environmentd`'s health. The tl;dr: this design doc proposes a new RPC service that lives in `envd`, that will broadcast persist state updates to any `clusterd` operating over those shards. We expect impact to be lightweight, but we don't want to DoS `envd`'s more critical operations and would love to have any input on how to minimize & verify impact.

**Comment by mjibson** [comment]
Putting this in envd is fine, especially since it seems like we could move it elsewhere later if needed? The thing we need to avoid is putting it in the main coordinator thread. My read of the doc is that's not where it would go: it'd be some unrelated thread/task running that doesn't interact with the coordinator at all. Is that right?

**Comment by pH14** [comment]
> Putting this in envd is fine, especially since it seems like we could move it elsewhere later if needed? The thing we need to avoid is putting it in the main coordinator thread. My read of the doc is that's not where it would go: it'd be some unrelated thread/task running that doesn't interact with the coordinator at all. Is that right?

Yes on both counts: we could move it elsewhere fairly easily if needed + we will run the service in a separate task outside of the coordinator

File: doc/developer/design/20230418_persist_state_pubsub.md [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+or a `SUBSCRIBE`, so the number of broadcasted messages would be proportional to the number of ongoing read operations,
+and not the total number of shards.
+
+`environmentd` is chosen to host the RPC service as it necessarily holds a handle to each shard (and therefore is always
+interested in all diffs), and to align with the existing topology of the Controller, rather than introduce the novel
+complexity of point-to-point connections between unrelated `clusterd` processes or a net-new dedicated process. With this
+implementation we would anticipate the end-to-end latency between a writer in one `clusterd` pushing an update to a reader
+in another `clusterd` receiving that update to be in the 1-10ms range.
+
+### Service Discovery
```
Comment (pH14) [inline doc/developer/design/20230418_persist_state_pubsub.md:152]: sorry to bother again @alex-hunt-materialize 🙈 would you mind a quick review on this section regarding the Kube changes?

File: doc/developer/design/20230418_persist_state_pubsub.md [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+
+`clusterd` will need to know the address of the `environmentd` process in order to connect to the PubSub server. Locally,
+this will be done through a standard `127.0.0.1:6879` port and passed in to `clusterd` via command line argument.
```
Comment (alex-hunt-materialize) [inline doc/developer/design/20230418_persist_state_pubsub.md:155]: I don't understand this sentence. What do you mean by "Locally", and what is done on `127.0.0.1:6879`?

Comment (pH14) [reply doc/developer/design/20230418_persist_state_pubsub.md:155]: ah, I meant when doing local development we can have `envd` bind the server to `127.0.0.1:6879` and point `clusterd` to that address

Comment (alex-hunt-materialize) [reply doc/developer/design/20230418_persist_state_pubsub.md:155]: Maybe we could rephrase this to make it clear that this is for local development? The term "locally" is unclear.

**Review by alex-hunt-materialize — APPROVED** [review]

File: doc/developer/design/20230418_persist_state_pubsub.md [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+    targetPort: 6879
+```
+
+`envd` will be informed of this service name via command line argument, e.g. `--persist-pubsub-service-name persist-pubsub-1`
+and pass it along to its `clusterd` param when set.
+
+We introduce this new headless service, rather than reusing the existing `environmentd` Service for a few reasons:
+
+* The `environmentd` Service is public / external-facing, while the pubsub port should be strictly internal
+* To decouple the idea the notion that `environmentd` is necessarily responsible for providing the PubSub server. In the
```
Comment (aljoscha) [inline doc/developer/design/20230418_persist_state_pubsub.md:189]:
```suggestion
* To decouple the notion that `environmentd` is necessarily responsible for providing the PubSub server. In the
```

File: doc/developer/design/20230418_persist_state_pubsub.md [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+
+We will introduce several metrics to understand the behavior of the change:
+
+* Number of updates pushed
+* Number of updates received that apply cleanly & uncleanly
+* How frequently `Listen`s need to fallback to polling
+* End-to-end latency of appending data to observing it
+* Timings of `environmentd`'s broadcasting / load
+
+CI will be used to ensure we do not introduce novel crashes or halts, but we anticipate the most interesting data to
```
Comment (pH14) [inline doc/developer/design/20230418_persist_state_pubsub.md:228]: @philip-stoev @def- would love any ideas you have about testing this change. We'll be writing some unit tests to cover the basic logic, but since this change will be introducing some new networking to `environmentd` and `clusterd`, in addition to giving us a new path for applying updates to persist state, perhaps we want to stress it more? I'm particularly concerned about ensuring the clients reconnect if the server restarts / making sure the code does not panic / that processes spin up successfully if the feature is disabled or params are incorrect

Comment (philip-stoev) [reply doc/developer/design/20230418_persist_state_pubsub.md:228]: Let us wrap our heads around this and will get back to you. At the very least, we need to start using multi-process sources/sinks in the CI (as opposed to multi-worker). 

re restarts/reconnects, we will come up with a separate mzcompose workflows as needed.

File: doc/developer/design/20230418_persist_state_pubsub.md (hunk 6 above)
Comment (benesch) [reply doc/developer/design/20230418_persist_state_pubsub.md:155]: s/Locally/When using the process orchestrator/?

File: doc/developer/design/20230418_persist_state_pubsub.md [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+    # ensure this service only routes to a specific generation of `envd`
+    materialize.cloud/envd-generation: <envd-generation>
+  type: ClusterIP
+  clusterIP: None
+  ports:
+  - name: pubsub
+    port: 6879
+    protocol: TCP
+    targetPort: 6879
+```
```
Comment (benesch) [inline doc/developer/design/20230418_persist_state_pubsub.md:181]: You'll need to teach cloudtest to do this too! Which, is probably fine, but points further in the direction of "we haven't gotten the abstraction boundary right between cloud infra, the database, and our testing infra."

File: doc/developer/design/20230418_persist_state_pubsub.md [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+
+`envd` will be informed of this service name via command line argument, e.g. `--persist-pubsub-service-name persist-pubsub-1`
+and pass it along to its `clusterd` param when set.
+
+We introduce this new headless service, rather than reusing the existing `environmentd` Service for a few reasons:
+
+* The `environmentd` Service is public / external-facing, while the pubsub port should be strictly internal
+* To decouple the notion that `environmentd` is necessarily responsible for providing the PubSub server. In the
+future we may wish to [replace it with a different implementation](#external-pubsubmessage-bus), or route the Service to
+point towards different pods.
```
Comment (benesch) [inline doc/developer/design/20230418_persist_state_pubsub.md:191]: This part was super helpful, thanks!

**Review by benesch — COMMENTED** [review]
Was only able to do a quick pass so far, but this was immensely clarifying for me, @pH14—the cloud interop section in particular! Thanks very much for writing it.

**Review by philip-stoev — APPROVED** [review]

File: doc/developer/design/20230418_persist_state_pubsub.md (hunk 8 above)
Comment (philip-stoev) [reply doc/developer/design/20230418_persist_state_pubsub.md:228]: https://github.com/MaterializeInc/materialize/issues/19246 . Please add any comments about testing or QA to the issue.

Comment (pH14) [reply doc/developer/design/20230418_persist_state_pubsub.md:229]: Sounds great, thank you @philip-stoev!

**Comment by pH14** [comment]
Uh... don't quite know why/how CLAassistant decided to pull everyone ever into this PR. Apologies for the unneeded ping 🙈
